pylops/waveeqprocessing/_kirchhoff_cuda.py

```diff
@@ -0,0 +1,519 @@
+        elif not self.travsrcrec:  # len(inputs) == 7:
+            x_d = cuda.to_device(inputs[0])
+            y_d = cuda.to_device(inputs[1])
+            nsnr_d = np.int32(inputs[2])
+            nt_d = np.int32(inputs[3])
+            ni_d = np.int32(inputs[4])
+            itrav_d = cuda.to_device(inputs[5])
+            travd_d = cuda.to_device(inputs[6])
+            self._trav_kirch_matvec_cuda[self.num_blocks, self.num_threads_per_blocks](x_d, y_d, nsnr_d, nt_d, ni_d,
+                                                                                       itrav_d, travd_d)
+
+        if not self.dynamic:
+            cuda.synchronize()
+            y = y_d.copy_to_host()
+        return y
+
+    def _rmatvec_call(self, *inputs):
+        """ Handle the adjoint operation call, dispatching to appropriate CUDA kernels.
+
+               This method selects the appropriate kernel to execute based on the computation flags,
+               and performs the adjoint operation (matrix-vector multiplication with the transpose).
+
+               Parameters
+               ----------
+               *inputs : list
+                   List of input parameters required by the kernels.
+
+               Returns
+               -------
+               y : ndarray
+                   Output data (image) after adjoint operation.
+        """
+        if self.dynamic and self.travsrcrec:
+            y = self._process_streams(inputs[0], "_rmatvec")
+        elif self.travsrcrec:  # len(inputs) == 9
+            x_d = cuda.to_device(inputs[0])
+            y_d = cuda.to_device(inputs[1])
+            ns_d = np.int32(inputs[2])
+            nr_d = np.int32(inputs[3])
+            nt_d = np.int32(inputs[4])
+            ni_d = np.int32(inputs[5])
+            dt_d = np.float32(inputs[6])
+            trav_srcs_d = cuda.to_device(inputs[7])
+            trav_recs_d = cuda.to_device(inputs[8])
+            self._travsrcrec_kirch_rmatvec_cuda[self.num_blocks, self.num_threads_per_blocks](x_d, y_d, ns_d, nr_d,
+                                                                                              nt_d, ni_d, dt_d,
+                                                                                              trav_srcs_d, trav_recs_d)
+        elif not self.travsrcrec:  # len(inputs) == 7:
+            x_d = cuda.to_device(inputs[0])
+            y_d = cuda.to_device(inputs[1])
+            nsnr_d = np.int32(inputs[2])
+            nt_d = np.int32(inputs[3])
+            ni_d = np.int32(inputs[4])
+            itrav_d = cuda.to_device(inputs[5])
+            travd_d = cuda.to_device(inputs[6])
+            self._trav_kirch_rmatvec_cuda[self.num_blocks, self.num_threads_per_blocks](x_d, y_d, nsnr_d, nt_d, ni_d,
+                                                                                        itrav_d, travd_d)
+
+        if not self.dynamic:
+            cuda.synchronize()
```

Comment:
Why is this needed since for `not self.dynamic` you use only the default stream?

Reply:
for the dynamic case, for each stream, I need to synchronize it with the main stream in a looped sequence. This synchronization happens within self._process_streams().

Reply:
Sure, but you have `if not self.dynamic:` so this will never be run in the dynamic case anyways?

Reply:
correct